demos/portal-embed-demo/public/js/portalog/portals-controller.js

```diff
@@ -0,0 +1,239 @@
+            const isFollowed = evt.target.classList.contains('followed');
+            this._changeFollowStatus(isFollowed);
+        });
+    }
+
+    /**
+     * Generate a transparent UI to control the embedded content via message
+     * @returns {HTMLDivElement}
+     */
+    _genPortalMessageUI() {
+        const controller = document.createElement('div');
+        const prev = document.createElement('div');
+        const play = document.createElement('div');
+        const next = document.createElement('div');
+
+        controller.classList.add('message-controller');
+        prev.classList.add('message-button');
+        play.classList.add('message-button');
+        next.classList.add('message-button');
+
+        prev.addEventListener('click', (evt) => {
+            this.portal.postMessage({ control: 'prev' }, this.origin);
+        });
+        play.addEventListener('click', (evt) => {
+            const isPlaying = play.getAttribute('playing');
+            if (isPlaying === 'true') {
+                this.portal.postMessage({ control: 'pause' }, this.origin);
+                play.setAttribute('playing', false);
+            } else {
+                this.portal.postMessage({ control: 'play' }, this.origin);
+                play.setAttribute('playing', true);
+            }
+        });
+        next.addEventListener('click', (evt) => {
+            this.portal.postMessage({ control: 'next' }, this.origin);
+        });
+
+        controller.appendChild(prev);
+        controller.appendChild(play);
+        controller.appendChild(next);
+
+        return controller;
+    }
+
+    /**
+     * Removing an element
+     * @param {HTMLElement} elm 
+     */
+    _remove(elm) {
+        if (elm) {
+            elm.parentNode.removeChild(elm);
+        }
+    }
+
+    /**
+     * Resetting root's style
+     */
+    _resetRoot() {
+        this.root.style.transition = '';
+        this.root.style.width = '100%';
+        this.root.style.paddingTop = '65%';
+        this.root.style.top = '0px';
+        this.root.style.left = '0px';
+    }
+
+    /**
+     * Follow and unfollow
+     * @param {boolean} isFollowed 
+     */
+    _changeFollowStatus(isFollowed) {
+        const follow = document.querySelector('#follow');
+        if (isFollowed) {
+            follow.classList.remove('followed');
+            follow.textContent = 'Follow';
+        } else {
+            follow.classList.add('followed');
+            follow.textContent = 'Following';
+        }
+    };
+
+}
+
+let portalPort = 3001;
+if(location.search) {
+    const matchResult = location.search.match(/portalport=(\d{4})/);
+    if(matchResult[1]){
+        portalPort = matchResult[1];
+    }
+}
+
+if('HTMLPortalElement' in window){
+    // Create instance
+    const portalsController = new PortalsController(
+        document.querySelector('#embed'),
+        `http://localhost:${portalPort}`,
+        '/',
+        170
```

Comment:
Agreed. Since it's only used in the animation function, I'll replace it like this
``` javascript
    /**
     * Animates the embedded content to smoothly activate it.
     */
    animateAndActivate() {
        // A Y poistion that needs to be animated to
        // TODO: retrieve this value from the portal via postMessage
        // in the future
        const TARGET_Y = 170;
```
I kind of wanted to loosely couple this value from the PortalController class itself since the position requirement is from the TTT Archive side and not PORTALOG. So, added a TODO comment that in the future, it would be good to retrieve this value from messages. WDTY?

Reply:
Might not be a bad idea, though it might be overkill for demo purposes. Happy to defer to your judgment.

Reply:
I'll keep the TODO and work on it next week as a refactoring task (and check then if that's necessary).